### PR TITLE
3711 refactor api v1 imports service

### LINKED ISSUE #3711: Refactor /api/v1/imports/service/*
- [X] /api/v1/imports/service/:id/token_valid
- [X] /api/v1/imports/service/:id/list_files
- [X] /api/v1/imports/service/:id/auth_url
- [X] /api/v1/imports/service/:id/validate_code/:code
- [X] /api/v1/imports/service/:id/oauth_callback

*What to test* | Dropbox | GDrive | Mailchimp
---------------- | ----------- | ---------- | -------------
Import          | ok | ok  | ok
Set Sync      |  ok | ok | ok
Sync now     |  ok | ok | ok
Automatic sync | ok | ok | ok
Revoke and try | ok | ok | ok
Not accept oauth | ok | ok | ok
Do nothing and try again | ok | ok | ok

`oauth_callback` enpoint requires nginx configuration, since it rewrites the request. This will be done by sending a modified service name: `service_name = "v1_1_#{service_name}"` if import services flag is set. In that case nginx will redirect to v1_1 endpoint.

`validate_code` is not currently used but will be kept for future usage.

### PR DESCRIPTION
**Pull request by juanignaciosl**
Ok, after staging testing (including callback endpoint testing and cleanup), @Kartones CR this, please. Closes #3711

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3849/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3851/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3852/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3853/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3854/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3856/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3857/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3860/))

File: app/models/carto/user.rb [hunk 1]
```diff
@@ -146,7 +148,17 @@ def default_basemap
...
   end
 
-  private
+  def oauth_for_service(service)
+    synchronization_oauths.where(service: service).first
+  end
 
+  def add_oauth(service, token)
+    synchronization_oauths.create(
+        service:  service,
```
Comment (Kartones) [inline app/models/carto/user.rb]: is user added automatically upon creation? As I'm missing it here (and in the `oauth_for_service` method above)

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cartodb-ui",
-  "version": "3.13.13",
+  "version": "3.13.13-ded03",
```
Comment (Kartones) [inline package.json]: careful with removing this once finished

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3927/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3929/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3932/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3934/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3935/))

File: services/datasources/lib/datasources/url/dropbox.rb [hunk 3]
```diff
@@ -75,7 +75,11 @@ def get_auth_url(use_callback_flow=true)
           else
             @auth_flow = DropboxOAuth2FlowNoRedirect.new(@app_key, @app_secret)
           end
-          @auth_flow.start(CALLBACK_STATE_DATA_PLACEHOLDER.sub('user', @user.username).sub('service', DATASOURCE_NAME))
+          service_name = DATASOURCE_NAME
```
Comment (Kartones) [inline services/datasources/lib/datasources/url/dropbox.rb]: mm why not extracting this to the parent class `CartoDB::Datasources::Base`? You'd avoid repeating on each datasource

Comment (juanignaciosl) [inline services/datasources/lib/datasources/url/dropbox.rb]: You're right, on it.

File: spec/requests/carto/api/layers_controller_spec.rb [hunk 4]
```diff
@@ -19,9 +19,9 @@
 
       # new controller
       scope :module => 'carto/api', :format => :json do
-        get    '(/user/:user_domain)(/u/:user_domain)/api/v1_1/users/:user_id/layers'     => 'layers#custom_layers_by_user',   as: :api_v1_users_layers_index
-        get    '(/user/:user_domain)(/u/:user_domain)/api/v1_1/maps/:map_id/layers'     => 'layers#layers_by_map',   as: :api_v1_maps_layers_index
-        get    '(/user/:user_domain)(/u/:user_domain)/api/v1_1/maps/:map_id/layers/:id' => 'layers#show',    as: :api_v1_maps_layers_show
+        get    '(/user/:user_domain)(/u/:user_domain)/api/v1/users/:user_id/layers'     => 'layers#custom_layers_by_user',   as: :api_v1_users_layers_index
```
Comment (Kartones) [inline spec/requests/carto/api/layers_controller_spec.rb]: oh, I recall @rafatower telling me to put v1_1 as we always call by `xxx_url` helper, so I left them with the v1_1

**Comment by Kartones** [comment]
:+1:

File: spec/requests/carto/api/layers_controller_spec.rb (hunk 4 above)
Comment (juanignaciosl) [inline spec/requests/carto/api/layers_controller_spec.rb]: I've changed them so we can't use explicit urls (we'd notice it works when it shouldn't), and it's more consistent with `as:` naming. I can undo if you prefer.

Comment (Kartones) [inline spec/requests/carto/api/layers_controller_spec.rb]: no no, it's ok, I think it's nice to have them like this

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3942/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3956/))

**Pull request merged**